FreeRTOS/FreeRTOS-Kernel
# Heap protect

Implement configENABLE_HEAP_PROTECTOR.

Description
-----------
* Setting `configENABLE_HEAP_PROTECTOR` to `1` obfuscates heap block pointers by XORing them with an application supplied canary value. This obfuscation helps to catch heap corruption should a heap buffer overflow occur. 
* This PR also adds heap bounds checking to heap_4 and heap_5.
* This PR also adds some additional integer underflow checks. 

Test Steps
-----------
Test harnesses to verify the effect of the change in the POSIX simulator are available here:
* https://github.com/oliverlavery/FreeRTOS/tree/heap-harness
* https://github.com/oliverlavery/FreeRTOS/tree/heap-harness-5

With `configENABLE_HEAP_PROTECTOR` set to 1, overwriting a free list block leads to an assert on an out of bounds heap pointer. I have also manually verified allocation, free, and corruption in GDB.

The POSIX simulator full demo works correctly with both heap_4.c and heap_5.c regardless of the `configENABLE_HEAP_PROTECTOR` setting.

Checklist:
----------

Related Issue
-----------
N/A

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
